library/core/src/ptr/const_ptr.rs

```diff
@@ -1694,7 +1694,7 @@ impl<T: ?Sized> PartialEq for *const T {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T: ?Sized> Eq for *const T {}
 
-// Comparison for pointers
+/// Pointers are compared as if they are usize
```

Comment:
```suggestion
/// Pointers are compared as an unsigned value with exposed provenance.
```

I think mentioning the provenance story may be worthwhile here, because aiui otherwise comparing pointers with different provenance would be unsound. @RalfJung is this accurate?

Could you add the same docs to `PartialEq` and `PartialOrd`? `*mut` should get similar changes https://github.com/rust-lang/rust/blob/aa4cfd0809064503c69795e8e83ad067ad9e11a1/library/core/src/ptr/mut_ptr.rs#L2100-L2129.

Reply:
Provenance does *not* get exposed by comparison so the "with exposed provenance" part doesn't seem right. I don't even know what exactly it should mean though.

Reply:
> I think mentioning the provenance story may be worthwhile here, because aiui otherwise comparing pointers with different provenance would be unsound

why would this comparison be unsound?

Reply:
I think the easiest way to document this is to just say that pointers are compared by comparing the result of calling `addr()`. That covers most everything relevant:
- Provenance does not get exposed
- Provenance does not influence the result

It does not say what happens with wide pointers, but that is an orthogonal question anyway.

Reply:
> Provenance does _not_ get exposed by comparison

I said the wrong thing here; I meant to say that provenance is not taken into account, not that it is exposed (sorry, provenance lingo is all still all a bit new to me...).

> why would this comparison be unsound?

AIUI any kind of comparison of pointers across allocations would ride the non-deterministic/unsound line if provenance matters, but this isn't a problem based on the above.

Ralf or Jubille if you want to yoink this review feel free, I'm happy to though if you have no preference.

Reply:
@tgross35 There's some nondeterminism always because, well, pointers are simply like that. Especially in const eval which can cause pointer comparisons to spuriously fail! That's why [`guaranteed_eq`](https://doc.rust-lang.org/std/primitive.pointer.html#method.guaranteed_eq-1) exists. But nondet != unsound.

afaik it is of course strictly unsound in C and C++, but we're not C or c++